README

This repository contains scripts to aid in setting up tests that
require a minimal functioning iRODS installation. (Minimal meaning
that the ICAT database server, iRODS server and disk resource all
reside on localhost.) The iRODS server will expect a pre-configured
PostgreSQL server to be available on localhost.

These scripts set up a clean, throwaway iRODS instance suitable for
running basic tests.

Versions of iRODS supported:

   3.3.1 (run-in-place) - Built and installed from git a repository.
>= 4.1.9 (packaged)     - Installed from the official RENCI FTP site.

A version 3.3.1 installation will be in local directory owned by the
current user. WARNING! The installation script supplied with iRODS
3.3.1 will make modifications to that user's $HOME/.odbc.ini. A
version >= 4.1.9 installation will be in the location defined by the
package.

In all cases the current user will have been given an active rodsamin
account, set up for passwordless login to iRODS using iinit.

As some systems may require more than the single default iRODS
resource to be available, an additional resource is added
automatically at install time. These are

   3.3.1 - a resource named 'testResc' with a vault at
           ${IRODS_RIP_DIR}/iRODS/TestVault (see below for the definition
           of 'IRODS_RIP_DIR').

>= 4.1.9 - a resource named 'testResc' with a vault at
           /var/lib/irods/iRODS/TestVault


The scripts supplied are:

./scripts/download_and_verify_irods.sh

    Downloads source or packages as appropriate and makes such
    verification checks as are available.

./scripts/install_irods.sh

    Builds source or installs packages.

./scripts/configure_irods.sh

    Configures an iRODS admin user with a default password for
    testing.


The script behaviour may be modified by setting environment variables.
The most important variable, relevant to both iRODS 3.3.1 and >= 4.1.9
is:

IRODS_VERSION        The iRODS version to install:
                     3.3.1, 4.1.10 (default)


Relevant to 3.3.1 run-in-place install only are:

IRODS_RIP_DIR        A directory to which the run-in-place install will
                     be relocated.


Relevant to >= 4.1.9 packaged install only are:

PG_PLUGIN_VERSION    The iRODS PostgeSQL plugin version to install:
                     1.10 (default)

PLATFORM             The RENCI name for the host platform:
                     ubuntu12 (default)

RENCI_FTP_URL        The RENCI FTP site:
                     ftp://ftp.renci.org (defaults now to https://dnap.cog.sanger.ac.uk)


Changes
Release 1.3

 - Use Sanger COG to provide iRODS packages.

 - Install libssl-dev which is required for building against iRODS.

 - Added iRODS 4.2.2

 - Bug fix: added missing dependencies (for non-Travis builds) and
   default PLATFORM

Release 1.2

 -  Make iRODS 4.1.10 the default. Support multiple iRODS 4.1.x versions.

 -  Use curl instead of wget because it allows to use a file:// URL to get
    packages.

Release 1.1

 -  iRODS 3.x may be relocated at install time by setting IRODS_RIP_DIR.


Release 1.0


scripts/install.sh
#!/bin/bash

# set -e -x
set -x

IRODS_VERSION=${IRODS_VERSION:=4.1.9}

install_common() {
    sudo apt-get install -qq odbc-postgresql unixodbc-dev
}

install_3_3_1() {
    cp ./config/odbc.ini $HOME/.odbc.ini

    vault_path=$PWD/irods-legacy/iRODS/Vault
    mkdir -p $vault_path

    sed -i.bak -e "s#__VAULT__#$vault_path#" ./config/irodssetup.in
    sed -i.bak -e 's/i386-linux-gnu/x86_64-linux-gnu/' ./irods-legacy/iRODS/scripts/perl/utils_platform.pl

    cd ./irods-legacy/iRODS
    ./irodssetup < ../../config/irodssetup.in
    ./irodsctl stop

    export CFLAGS=-fPIC
    make
}

install_4_1_x() {
    sudo apt-get install -qq python-psutil python-requests
    sudo apt-get install -qq python-sphinx
    sudo apt-get install super libjson-perl jq
    sudo -H pip install jsonschema

    sudo dpkg -i irods-icat-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb irods-database-plugin-postgres-${PG_PLUGIN_VERSION}-${PLATFORM}-${ARCH}.deb
    sudo dpkg -i irods-runtime-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb irods-dev-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
}

case $IRODS_VERSION in

    3.3.1)
        install_common
        install_3_3_1
        ;;

    4.1.9)
        install_common
        install_4_1_x
        ;;

    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac


scripts/configure.sh
#!/bin/bash

set -e -x

IRODS_VERSION=${IRODS_VERSION:=4.1.9}

configure_common() {
    return
}

configure_3_3_1() {
    # The iRODS 3.3.1 setup script has created the irods user, the
    # irods database user and ICAT database already.
    cd ./irods-legacy/iRODS
    ./irodsctl restart

    export PATH=$PWD/clients/icommands/bin:$PATH

    test_resource=testResc
    test_vault=./irods-legacy/iRODS/Test

    mkdir -p $test_vault
    iadmin mkresc testResc 'unix file system' cache `hostname --fqdn` $test_vault
    iadmin asq 'select alias,sqlStr from R_SPECIFIC_QUERY where alias = ?' findQueryByAlias
}

configure_4_1_x() {
    sudo -E -u postgres createuser -D -R -S irods
    sudo -E -u postgres createdb -O irods ICAT
    sudo -E -u postgres sh -c "echo \"ALTER USER irods WITH PASSWORD 'irods'\" | psql"

    sudo /var/lib/irods/packaging/setup_irods.sh < ./config/setup_irods.sh.in
    sudo jq -f ./config/server_config.delta /etc/irods/server_config.json > server_config.tmp
    sudo mv server_config.tmp /etc/irods/server_config.json
    sudo /etc/init.d/irods restart

    test_resource=testResc
    test_vault=/var/lib/irods/iRODS/Test
    test_password=testpass

    sudo -E su irods -c "mkdir -p $test_vault"
    sudo -E su irods -c "iadmin mkresc unixfs unixfilesystem `hostname --fqdn`:$test_vault"

    sudo -E su irods -c "iadmin mkresc $test_resource replication"
    sudo -E su irods -c "iadmin addchildtoresc $test_resource unixfs"

    sudo -E su irods -c "iadmin mkuser $USER rodsadmin"
    sudo -E su irods -c "iadmin moduser $USER password $test_password"

    mkdir $HOME/.irods
    sed -e "s#__USER__#$USER#" -e "s#__HOME__#$HOME#" < ./config/irods_environment.json > $HOME/.irods/irods_environment.json
    cat $HOME/.irods/irods_environment.json

    echo $test_password | script -q -c "iinit"
}

case $IRODS_VERSION in

    3.3.1)
        configure_common
        configure_3_3_1
        ;;

    4.1.9)
        configure_common
        configure_4_1_x
        ;;

    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac


scripts/install_irods.sh
#!/bin/bash

set -e -x

BUILD_DIR=${BUILD_DIR:=$PWD}

ARCH=${ARCH:=x86_64}
PLATFORM=${PLATFORM:=ubuntu12}

IRODS_VERSION=${IRODS_VERSION:=4.1.10}
PG_PLUGIN_VERSION=${PG_PLUGIN_VERSION:=1.10}
IRODS_RIP_DIR=${IRODS_RIP_DIR:=/usr/local/irods}

PGHOME=${PGHOME:=/usr/lib/postgresql}
PGVERSION=${PGVERSION:=9.3}

install_common() {
    sudo apt-get install -y -q postgresql-client odbc-postgresql unixodbc-dev
}

install_3_3_1() {
    cp ${BUILD_DIR}/config/odbc.ini $HOME/.odbc.ini

    postgres_home=${PGHOME}/${PGVERSION}
    vault_path=${IRODS_RIP_DIR}/iRODS/Vault
    mkdir -p $vault_path
    chown -R $USER:$USER $vault_path

    sed -i.bak -e "s#__POSTGRES_HOME__#$postgres_home#" ${BUILD_DIR}/config/irodssetup.in
    sed -i.bak -e "s#__VAULT__#$vault_path#" ${BUILD_DIR}/config/irodssetup.in
    sed -i.bak -e 's/i386-linux-gnu/x86_64-linux-gnu/' ${IRODS_RIP_DIR}/iRODS/scripts/perl/utils_platform.pl

    cd ${IRODS_RIP_DIR}/iRODS
    ./irodssetup < ${BUILD_DIR}/config/irodssetup.in
    ./irodsctl stop

    # Rebuild with -fPIC for libRodsAPI.a
    export CFLAGS=-fPIC
    make clean
    make
}

install_4_1_x() {
    sudo apt-get install -y -q libssl-dev
    sudo apt-get install -y -q python-pip python-psutil python-requests
    sudo apt-get install -y -q python-sphinx
    sudo apt-get install -y -q super libjson-perl jq
    sudo -H pip install jsonschema

    sudo dpkg -i irods-icat-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb irods-database-plugin-postgres-${PG_PLUGIN_VERSION}-${PLATFORM}-${ARCH}.deb
    sudo dpkg -i irods-runtime-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb irods-dev-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
}

install_4_2_x() {
    sudo apt-get install -y -q irods-server irods-database-plugin-postgres
    sudo apt-get install -y -q irods-icommands
    sudo apt-get install -y -q irods-dev
    sudo apt-get install -y -q libssl-dev
}

case $IRODS_VERSION in

    3.3.1)
        install_common
        install_3_3_1
        ;;

    4.1.*)
        install_common
        install_4_1_x
        ;;

    4.2.*)
        install_common
        install_4_2_x
        ;;
    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac


scripts/configure_irods.sh
#!/bin/bash

set -e -x

BUILD_DIR=${BUILD_DIR:=$PWD}

IRODS_VERSION=${IRODS_VERSION:=4.1.10}
IRODS_RIP_DIR=${IRODS_RIP_DIR:=/usr/local/irods}

configure_common() {
    return
}

configure_3_3_1() {
    # The iRODS 3.3.1 setup script has created the irods user, the
    # irods database user and ICAT database already.
    cd ${IRODS_RIP_DIR}/iRODS
    ./irodsctl restart

    export PATH=${IRODS_RIP_DIR}/iRODS/clients/icommands/bin:$PATH

    test_resource=testResc
    test_vault=${IRODS_RIP_DIR}/iRODS/TestVault

    sudo mkdir -p $test_vault
    sudo chown -R $USER:$USER $test_vault

    iadmin mkresc testResc 'unix file system' cache `hostname --fqdn` $test_vault
    iadmin asq 'select alias,sqlStr from R_SPECIFIC_QUERY where alias = ?' findQueryByAlias
}

configure_postgres_4_x_x() {
    sudo -E -u postgres createuser -D -R -S irods
    sudo -E -u postgres createdb -O irods ICAT
    sudo -E -u postgres sh -c "echo \"ALTER USER irods WITH PASSWORD 'irods'\" | psql"
}

configure_checksum_4_x_x() {
    sudo jq -f ${BUILD_DIR}/config/server_config.delta /etc/irods/server_config.json > server_config.tmp
    sudo mv server_config.tmp /etc/irods/server_config.json
    sudo /etc/init.d/irods restart
}

configure_resources_4_x_x() {
    test_resource=testResc
    test_vault=/var/lib/irods/iRODS/TestVault
    test_password=testpass

    sudo -E su irods -c "mkdir -p $test_vault"
    sudo -E su irods -c "iadmin mkresc unixfs unixfilesystem `hostname --fqdn`:$test_vault"

    sudo -E su irods -c "iadmin mkresc $test_resource replication"
    sudo -E su irods -c "iadmin addchildtoresc $test_resource unixfs"

    sudo -E su irods -c "iadmin mkuser $USER rodsadmin"
    sudo -E su irods -c "iadmin moduser $USER password $test_password"
}

configure_login_4_x_x() {
    mkdir $HOME/.irods
    sed -e "s#__USER__#$USER#" -e "s#__HOME__#$HOME#" < ${BUILD_DIR}/config/irods_environment.json > $HOME/.irods/irods_environment.json
    cat $HOME/.irods/irods_environment.json

    echo $test_password | script -q -c "iinit" /dev/null
}

configure_4_1_x() {
    configure_postgres_4_x_x

    sudo /var/lib/irods/packaging/setup_irods.sh < ${BUILD_DIR}/config/setup_irods.sh.in

    configure_checksum_4_x_x
    configure_resources_4_x_x
    configure_login_4_x_x
}

configure_4_2_x() {
    configure_postgres_4_x_x

    sudo python /var/lib/irods/scripts/setup_irods.py < ${BUILD_DIR}/config/setup_irods.py.in

    configure_checksum_4_x_x
    configure_resources_4_x_x
    configure_login_4_x_x
}

case $IRODS_VERSION in

    3.3.1)
        configure_common
        configure_3_3_1
        ;;

    4.1.*)
        configure_common
        configure_4_1_x
        ;;

    4.2.*)
        configure_common
        configure_4_2_x
        ;;

    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac


scripts/download_and_verify.sh
#!/bin/bash

set -e -x

IRODS_VERSION=${IRODS_VERSION:=4.1.9}
PLATFORM=${PLATFORM:=ubuntu12}
ARCH=x86_64

RENCI_FTP_URL=${RENCI_FTP_URL:=ftp://ftp.renci.org}
WTSI_NPG_GITHUB_URL=${WTSI_NPG_GITHUB_URL:=https://github.com/wtsi-npg}

before_install_common() {
    echo sudo apt-get update -qq
}

before_install_3_3_1() {
    git clone ${WTSI_NPG_GITHUB_URL}/irods-legacy-gclp irods-legacy
}

before_install_4_1_x() {
    wget ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-database-plugin-postgres-${PG_PLUGIN_VERSION}-${PLATFORM}-${ARCH}.deb
    wget ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-dev-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
    wget ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-icat-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
    wget ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-runtime-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb

    sha256sum -c ./checksums/packages.sha256
}

case $IRODS_VERSION in

    3.3.1)
        before_install_common
        before_install_3_3_1
        ;;

    4.1.9)
        before_install_common
        before_install_4_1_x
        ;;

    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac


scripts/download_and_verify_irods.sh
#!/bin/bash

set -e -x -o pipefail

BUILD_DIR=${BUILD_DIR:=$PWD}

ARCH=${ARCH:=x86_64}
PLATFORM=${PLATFORM:=ubuntu12}

IRODS_VERSION=${IRODS_VERSION:=4.1.10}
PG_PLUGIN_VERSION=${PG_PLUGIN_VERSION:=1.10}
IRODS_RIP_DIR=${IRODS_RIP_DIR:=/usr/local/irods}

RENCI_FTP_URL=${RENCI_FTP_URL:=https://dnap.cog.sanger.ac.uk}
RENCI_PKG_URL=${RENCI_PKG_URL:=https://packages.irods.org}
WTSI_NPG_GITHUB_URL=${WTSI_NPG_GITHUB_URL:=https://github.com/wtsi-npg}
WTSI_NPG_GITHUB_REPO=${WTSI_NPG_GITHUB_REPO:=irods-legacy-gclp}

before_install_common() {
    sudo apt-get update
}

before_install_3_3_1() {
    sudo mkdir -p ${IRODS_RIP_DIR}
    sudo chown -R $USER:$USER ${IRODS_RIP_DIR}
    git clone ${WTSI_NPG_GITHUB_URL}/${WTSI_NPG_GITHUB_REPO} ${IRODS_RIP_DIR}
}

before_install_4_1_x() {
    curl -sSL -O ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-database-plugin-postgres-${PG_PLUGIN_VERSION}-${PLATFORM}-${ARCH}.deb
    curl -sSL -O ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-dev-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
    curl -sSL -O ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-icat-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb
    curl -sSL -O ${RENCI_FTP_URL}/pub/irods/releases/${IRODS_VERSION}/${PLATFORM}/irods-runtime-${IRODS_VERSION}-${PLATFORM}-${ARCH}.deb

    sha256sum -c ${BUILD_DIR}/checksums/packages-${IRODS_VERSION}.sha256
}

before_install_4_2_x() {
    curl -sSL -O ${RENCI_PKG_URL}/irods-signing-key.asc
    if [ $(grep -c 'BEGIN PGP PUBLIC KEY BLOCK' irods-signing-key.asc) != 1 ];
    then
        echo 'More than one GPG key in irods-signing-key.asc'
        exit 1
    fi

    sudo apt-key add irods-signing-key.asc
    echo "deb [arch=amd64] ${RENCI_PKG_URL}/apt/ $(lsb_release -sc) main" | sudo tee /etc/apt/sources.list.d/renci-irods.list
    sudo apt-get update
}

case $IRODS_VERSION in

    3.3.1)
        before_install_common
        before_install_3_3_1
        ;;

    4.1.*)
        before_install_common
        before_install_4_1_x
        ;;

    4.2.*)
        before_install_common
        before_install_4_2_x
        ;;
    *)
        echo Unknown iRODS version $IRODS_VERSION
        exit 1
esac
